src/core/PhaseVocoder.cpp

```diff
@@ -0,0 +1,272 @@
+/*
+ * PhaseVocoder.cpp - Implementation of the PhaseVocoder class
+ *
+ * Copyright (c) 2023 Daniel Kauss Serna <daniel.kauss.serna@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+#include "PhaseVocoder.h"
+
+#include <cstdio>
+#include <fftw3.h>
+#include <math.h>
+#include <typeinfo>
+
+#include "lmms_basics.h"
+#include "lmms_constants.h"
+
+namespace lmms {
+
+PhaseVocoder::PhaseVocoder()
+	: m_FFTInput(s_windowSize, 0)
+	, m_IFFTReconstruction(s_windowSize, 0)
+	, m_allMagnitudes(s_windowSize, 0)
+	, m_allFrequencies(s_windowSize, 0)
+	, m_processedFreq(s_windowSize, 0)
+	, m_processedMagn(s_windowSize, 0)
+{
+	m_fftPlan = fftwf_plan_dft_r2c_1d(s_windowSize, m_FFTInput.data(), m_FFTSpectrum.data(), FFTW_MEASURE);
+	m_ifftPlan = fftwf_plan_dft_c2r_1d(s_windowSize, m_FFTSpectrum.data(), m_IFFTReconstruction.data(), FFTW_MEASURE);
+}
+
+PhaseVocoder::~PhaseVocoder()
+{
+	fftwf_destroy_plan(m_fftPlan);
+	fftwf_destroy_plan(m_ifftPlan);
+}
+
+void PhaseVocoder::loadData(std::vector<float> originalData, int sampleRate, float newRatio)
+{
+	m_dataLock.lock();
+
+	m_originalBuffer = originalData;
```

Comment:
```suggestion
	m_originalBuffer = std::move(originalData);
```
There are 2 extra copies of the original data. Making the `originalData` parameter a const reference would fix one of them, but the second is at the `loadSample()` call site and requires moving the 2 vectors produced there into this method.

Reply:
Sorry I don't think I'm following. The data gets loaded in `SlicerT::updataFile`, and then a pointer gets passed to `DynamicPlaybackBuffer::loadSample`. This only creates one copy, the one in `updateFile`, since the `DynamicPlaybackBuffer` stores a pointer. Afterwards, `DynamicPlaybackBuffer` de interlaces the pointer into left and right, and then creates a copy to pass them to `PhaseVocoder::loadData`. I only see one extra copy here, the one that gets created for `PhaseVocoder:loadData`. This copy can be avoided, like you said, by passing a const reference. So what is the move for then? And what does [this review](https://github.com/LMMS/lmms/pull/6857#discussion_r1359924843) fix exactly?

Reply:
Originally, you had this:
`DynamicPlaybackBuffer::loadSample(...)` creates two `std::vector<float>`s and passes them to `PhaseVocoder::loadData(...)` by value. Passing by value creates one unnecessary copy of each vector.

Then in `PhaseVocoder::loadData(...)`, you took the vector and assigned it to `m_originalBuffer`. This creates a second unnecessary copy of each vector.

By changing `PhaseVocoder::loadData(...)` to use a const reference parameter, you fixed the first unnecessary copy, but not the second, since assigning a `const std::vector<float>&` to `m_originalBuffer` still copies the vector. The vectors from `loadSample` need to be moved into `loadData` then moved again into `m_originalBuffer` if you want to avoid any unnecessary copies.

To do that, I'd recommend a `PhaseVocoder::loadData(...)` method signature that takes the vector by value, and moving the vector in the body of that method. Then, when calling the method, you'd use `std::move` to avoid both unnecessary copies.

Reply:
Done, although I don't know if this will have any real effect.